Refactoring

BjornMelgaard opened the pull request:
@Sly777 , I took the liberty of *slight* refactoring, would you accept it? (based on scalable-react-typescript-boilerplate)

P.S. Im on a middle of refactoring

Sly777 commented:
Hey @BjornMelgaard, Thanks for the massive change! :)

My opinions:
- I don't like this folder structure honestly (components/layouts/containers). It's making complexity unnecessarily. Components/Containers are really enough.
- File structure also can stay same for the component (index.js, index.data.js, index.style.js) 
- Splitting style from component file is good 👍 
- eslint-watch is okay but I think we can find the better way for it (integration with webpack & next.js)
- I was thinking same for redux dev-tool. 👍

BjornMelgaard commented:
hi, @Sly777 , I made file structure as you said (not index.data.js, but data.js), removed redux-dev-tools, changed validation code, so it works now

[there small issue, you can participate)))](https://stackoverflow.com/questions/45610529/styled-components-parent-styles-has-more-priority-then-child)

BjornMelgaard commented:
Lovely, but problem with logout button left

Sly777 commented on libraries/middleware.js:
```diff
@@ -1,9 +1,11 @@
-  const composeEnhancers = composeWithDevTools({});
+  const composeEnhancers =
+    (typeof window !== 'undefined' &&
+      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
```
I think It should be ```window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})```. I'm not perfectly sure.

Sly777 commented on pages/signup.js:
```diff
@@ -1,9 +1,9 @@
-import SignUpForm from '../components/SignUpForm';
+import SignUpForm from '../containers/SignUpForm';
```
SignupForm is not the container. It's just component.

BjornMelgaard replied:
@Sly777, thanks for review
I will make as before
Can you explain how you divide between containers and components?

Sly777 replied:
Containers are basically used by pages. If you have a function (or class) that has children and also containing some other components that are same on the other pages and also used by pages one time only (not like a list), That means it's container. But If you have a function that has only props and can be used by the other components, containers and pages, that means it's component.

BjornMelgaard replied:
Components dumb, containers smart
If component is using data from apollo or redux - its container
SignUpForm isnt component, its container
All components here is AuthFields and LinkList

**lets war begin!!!**

Sly777 commented on pages/signup.js:
```diff
@@ -1,9 +1,9 @@
+import SignUpForm from '../containers/SignUpForm';
 import withData from '../libraries/withData';
-import DefaultCon from '../containers/Default';
+import Layout from '../containers/Layout';
```
It should stay with the same name Because you can understand what it's ```Default``` container. ```Layout``` doesn't have same meaning.

Sly777 commented on libraries/reduxStore.js:
```diff
@@ -2,7 +2,7 @@ import { createStore } from 'redux';
 import createMiddleware from './middleware';
 import persist from './persist';
-import { dispatchers } from '../components/AuthFields/index.data';
+import { dispatchers } from '../containers/AuthFields/store';
```
It should stay in components folder.

Sly777 commented on package.json:
```diff
@@ -8,7 +8,8 @@
     "start:next": "next start",
-    "lint": "eslint --fix --ext .gql --ext .graphql --ext .js pages components libraries server helper_scripts",
+    "lint": "eslint --fix --ext .gql --ext .graphql --ext .js pages components containers layouts libraries server helper_scripts",
+    "lint:watch": "esw -w --fix --ext .gql --ext .graphql --ext .js pages components containers layouts libraries server helper_scripts",
```
I will search other better options for this. Because there is hot reload on next.js side.

Sly777 replied:
We can try this
https://github.com/zeit/next.js/issues/2184

BjornMelgaard replied:
very interesting)

BjornMelgaard replied:
@Sly777 eslint-loader cant watch and fix, as I remenber (not shure)

BjornMelgaard replied:
I'll try

Sly777 replied:
:) Yeah, it cannot watch but next.js does. So when next.js detected changes, it will run also lint on the webpack side.

BjornMelgaard replied:
```
    if (dev) {
      config.module.rules.push({
        test: /\.(jsx?|gql|graphql)$/,
        loader: 'eslint-loader',
        exclude: /node_modules/,
        enforce: 'pre',
        options: {
          fix: true
        }
      })
    }

```

with this js files are checked and fixed, but not gql

BjornMelgaard replied:
and server too (but you know this) )))
maybe left eslint-watch?

Sly777 commented on containers/SignInForm/index.js:
```diff
@@ -56,20 +49,27 @@ class SignInForm extends React.Component {
       .then(response => {
         if (response.data) {
-          this.props.signinUserDispatcher(response.data.signinUser.token);
+          this.props.actions.signInUser(response.data.signinUser.token);
```
We can say ```signIn``` for this

Sly777 commented on containers/AuthFields/store.js:
```diff
@@ -24,30 +24,28 @@ const reducer = (state = initialState, action) => {
       };
     case AUTH_SIGNOUT:
       return { ...state, authenticated: false, token: null, error: null };
-    case AUTH_SERVERERROR:
```
We can use this.

BjornMelgaard commented:
tomorrow will try storybook, yay

BjornMelgaard commented:
I have fixed issue with Logout button and believe it can be merged

Sly777 commented:
Hey @BjornMelgaard, Everything looks good but please don't add new feature inside of it anymore because It's started to be uncheckable and messy code. I really want to merge this because there are really effective and good changes by you but right now, It's been really big PR. My opinions are:

- Please create another PRs for ```babel-plugin-root-import```, ```jest``` and ```eslint automatically``` features.
- It can be better if don't change the ```component-container``` structure. We can talk about it later on another PR.

BjornMelgaard commented:
@Sly777 , sorry about that, I got carried away 

Can we decide right now about `component-container`? (I'm tired of remaking) 
I really think this is clear and better separation: If component is using data from apollo or redux - its container. [scalable-react-typescript-boilerplate](https://github.com/scalable-react/scalable-react-typescript-boilerplate/tree/master/packages/docs/src/client) is using this separation 

P.S.
though they are wrong in locating store code (reducers, actions,...) in same directory with containers: if you have `SignUp` and `SignIn` containers in which of them locate `auth store`?

P.P.S.
In `libraries/stores`

Sly777 commented:
Sure, I have some concerns but let's talk @BjornMelgaard and find the same way :) but it shouldn't be in this PR :D Firstly, let's merge this and then we can talk on another pr.

BjornMelgaard commented:
done

BjornMelgaard commented:
@Sly777 you have two eslint-plugin-react
